### PR TITLE
Allow anonymous template types

### LINKED ISSUE #574: assertion failed: ctx.resolve_type_through_type_refs(decl).is_opaque()
### Input C/C++ Header

```C++
template <class> class a {};
class {
  a<int> ar;
} AutoIdVector;
```

### Bindgen Invokation

```
$ bindgen input.hpp -- -std=c++14
```

### Actual Results

```
thread 'main' panicked at 'assertion failed: ctx.resolve_type_through_type_refs(decl).is_opaque()', /home/fitzgen/rust-bindgen/target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4636
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: <bindgen::ir::template::TemplateInstantiation as bindgen::codegen::ToRustTy>::to_rust_ty
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4636
   1: <bindgen::ir::ty::Type as bindgen::codegen::ToRustTy>::to_rust_ty
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4474
   2: <bindgen::ir::item::Item as bindgen::codegen::ItemToRustTy>::to_rust_ty
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4309
   3: <bindgen::ir::context::ItemId as bindgen::codegen::ItemToRustTy>::to_rust_ty
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4304
   4: <bindgen::ir::ty::Type as bindgen::codegen::ToRustTy>::to_rust_ty
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4476
   5: <bindgen::ir::item::Item as bindgen::codegen::ItemToRustTy>::to_rust_ty
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4309
   6: <bindgen::ir::context::ItemId as bindgen::codegen::ItemToRustTy>::to_rust_ty
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4304
   7: <bindgen::ir::comp::CompInfo as bindgen::codegen::CodeGenerator>::codegen
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:2581
   8: <bindgen::ir::ty::Type as bindgen::codegen::CodeGenerator>::codegen
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:1508
   9: <bindgen::ir::item::Item as bindgen::codegen::CodeGenerator>::codegen
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:1301
  10: <bindgen::ir::module::Module as bindgen::codegen::CodeGenerator>::codegen::{{closure}}
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:1322
  11: <bindgen::ir::module::Module as bindgen::codegen::CodeGenerator>::codegen
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:1346
  12: <bindgen::ir::item::Item as bindgen::codegen::CodeGenerator>::codegen
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:1287
  13: bindgen::codegen::codegen::{{closure}}
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4955
  14: bindgen::ir::context::BindgenContext::gen
             at ./src/ir/context.rs:578
  15: bindgen::codegen::codegen
             at ./target/debug/build/bindgen-fddff672ca36ebc4/out/codegen.rs:4933
  16: bindgen::Bindings::generate
             at ./src/lib.rs:748
  17: bindgen::Builder::generate
             at ./src/lib.rs:469
  18: bindgen::main::{{closure}}
             at ./src/main.rs:55
  19: std::panicking::try::do_call
             at /checkout/src/libstd/panicking.rs:454
  20: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  21: std::panicking::try
             at /checkout/src/libstd/panicking.rs:433
  22: std::panic::catch_unwind
             at /checkout/src/libstd/panic.rs:361
  23: bindgen::main
             at ./src/main.rs:54
```

### Expected Results

No panic, bindings generated.

### PR DESCRIPTION
**Pull request by fitzgen**
We have various assertions that the only way that some template parameter related methods will return `None` is if the template definition is marked opaque. These assertions fail in the presence of test cases with unnamed template types, because we never save an IR item for the template type, and subsequently think that the template definition has no template parameters. The assertions are in fact sound and correct, so it doesn't make sense to remove them. Instead it is more correct to save IR items for the anonymous template types and simply let the template usage analysis prevent them from getting
codegen'd.

Fixes #574

r? @emilio

### CODE REVIEW

**Comment by fitzgen** [comment]
See the first commit message

File: src/ir/comp.rs [hunk 1]
```diff
@@ -357,8 +359,8 @@ impl CompInfo {
 
         let has_destructor = self.has_destructor ||
                              match self.kind {
-            CompKind::Union => false,
-            CompKind::Struct => {
+                                 CompKind::Union => false,
```
Comment (emilio) [inline src/ir/comp.rs]: I don't think these changes are really acceptable in general.

Is this only due to a newer version of rustfmt?

File: src/ir/module.rs [hunk 2]
```diff
@@ -63,7 +63,7 @@ impl DotAttributes for Module {
                          _ctx: &BindgenContext,
                          out: &mut W)
                          -> io::Result<()>
-        where W: io::Write,
+        where W: io::Write
```
Comment (emilio) [inline src/ir/module.rs]: Also, it's no longer respecting the trailing comma in where clauses?

File: src/ir/named.rs [hunk 3]
```diff
@@ -326,10 +326,10 @@ impl<'ctx, 'gen> MonotoneFramework for UsedTemplateParameters<'ctx, 'gen> {
                 // between nodes.
                 item.trace(ctx,
                            &mut |sub_item, _| {
-                               dependencies.entry(sub_item)
-                                   .or_insert(vec![])
-                                   .push(item);
-                           },
+                                    dependencies.entry(sub_item)
```
Comment (emilio) [inline src/ir/named.rs]: This change seems also wrong to me.

File: src/main.rs [hunk 4]
```diff
@@ -52,8 +52,8 @@ pub fn main() {
         Ok((builder, output, verbose)) => {
 
             let builder_result = panic::catch_unwind(|| {
-                builder.generate().expect("Unable to generate bindings")
-            });
+                                                         builder.generate().expect("Unable to generate bindings")
```
Comment (emilio) [inline src/main.rs]: This is also way off I believe.

**Comment by fitzgen** [comment]
Thanks for the review! Dropped the last commit. Yeah, this is latest `rustfmt` don't know what's up...

@bors-servo r+

**Comment by fitzgen** [comment]
FWIW, filed https://github.com/rust-lang-nursery/rustfmt/issues/1376 for the `rustfmt` issues.

**Pull request merged**